crates/owl-ms-language-server/src/lib.rs

```diff
@@ -0,0 +1,869 @@
+mod catalog;
+pub mod debugging;
+mod position;
+mod queries;
+mod range;
+pub mod rope_provider;
+#[cfg(test)]
+mod tests;
+mod workspace;
+
+use core::panic;
+use debugging::timeit;
+use itertools::Itertools;
+use log::{debug, info};
+use once_cell::sync::Lazy;
+use position::Position;
+use queries::{ALL_QUERIES, NODE_TYPES};
+use range::Range;
+use rope_provider::RopeProvider;
+use std::collections::HashMap;
+use std::fs;
+use std::path::Path;
+use tokio::sync::{MappedMutexGuard, Mutex, MutexGuard};
+use tokio::task::{self};
+use tower_lsp::jsonrpc::Result;
+use tower_lsp::lsp_types::{self, *};
+use tower_lsp::{Client, LanguageServer};
+use tree_sitter::{InputEdit, Language, Node, Parser, Point, Query, QueryCursor, Tree, TreeCursor};
+use tree_sitter_owl_ms::language;
+use workspace::{gen_diagnostics, node_text, Document, Workspace};
+
+// Constants
+
+pub static LANGUAGE: Lazy<Language> = Lazy::new(language);
+
+// Model
+pub struct Backend {
+    pub client: Client,
+    parser: Mutex<Parser>, // stateful because of resume behavior after fail, timeout or cancellation
+    position_encoding: Mutex<PositionEncodingKind>,
+    workspaces: Mutex<Vec<Workspace>>,
+}
+
+impl Backend {
+    /// Creates a new [`Backend`].
+    pub fn new(client: Client, parser: Parser) -> Self {
+        Backend {
+            client,
+            parser: Mutex::new(parser),
+            position_encoding: PositionEncodingKind::UTF16.into(),
+            workspaces: Mutex::new(vec![]),
+        }
+    }
+
+    async fn lock_workspaces(&self) -> MappedMutexGuard<Vec<Workspace>> {
+        MutexGuard::map(self.workspaces.lock().await, |w| w)
```

Comment:
Is there a need for this function? You only use it in tests; in all other cases you directly lock `workspaces`. If yes, why use the `MappedMutexGuard` if the mapping function is the identity function?

Reply:
If yes, would it also make sense to move it to the second `impl` block? That would separate the public creation functions from internal utility functions.

Reply:
I think its not needed in release. The purpose is to lock all workspaces at once. This was needed because I cant move ownership of workspaces in tests to assert stuff. I should move it in some sort of testing utility module.